File: 2022/Day6/python/solve.py
import os


def solve(line, min_different):
    for i, _ in enumerate(line):
        if i < min_different - 1: continue

        chars = line[i:i+min_different]
        if len(set(chars)) == min_different:
            return i + min_different

dirname = os.path.dirname(__file__)
filename = os.path.join(dirname, '../input.txt')
with open(filename) as f:
    line = f.readline()

print(solve(line, 4))
print(solve(line, 14))

File: 2023/Day4/javascript/solve.js
const fs = require("node:fs");
const path = require("node:path");

const input = fs
  .readFileSync(path.join(__dirname, "../input.txt"), "utf8")
  .split("\n")
  .filter((line) => line !== "");

const part1 = (cards) => {
  const results = [];
  for (const card of cards) {
    const [_, winnindNumbersString, drawnNumbersString] =
      /Card +\d+: ([ \d]+) +\| +([ \d]+)/.exec(card);
    const winnindNumbers = winnindNumbersString
      .trim()
      .split(" ")
      .map((num) => Number.parseInt(num, 10))
      .filter((num) => !isNaN(num));

    const drawnNumbers = drawnNumbersString
      .trim()
      .split(" ")
      .map((num) => Number.parseInt(num, 10))
      .filter((num) => !isNaN(num));

    let matches = [];

    for (const num of drawnNumbers) {
      if (winnindNumbers.includes(num)) {
        matches.push(num);
      }
    }

    matches.length > 0 && results.push(2 ** (matches.length - 1));
  }

  return results.reduce((sum, curr) => sum + curr, 0);
};

console.log(part1(input));
